crates/circuit/src/circuit_data.rs

```diff
@@ -1038,13 +1296,76 @@ impl CircuitData {
             {
                 return Ok(false);
             }
+            let other_cd = other_cd.borrow();
+
+            if self_cd.num_input_vars() != other_cd.num_input_vars()
+                || self_cd.num_captured_vars() != other_cd.num_captured_vars()
+                || self_cd.num_declared_vars() != other_cd.num_declared_vars()
+                || self_cd.num_captured_stretches() != other_cd.num_captured_stretches()
+                || self_cd.num_declared_stretches() != other_cd.num_declared_stretches()
+            {
+                return Ok(false);
+            }
+
+            let mut prev_rhs_stretch_idx = 0usize;
+            for (id_name, lhs_id_info) in &self_cd.identifier_info {
+                let Some(rhs_id_info) = other_cd.identifier_info.get(id_name) else {
+                    return Ok(false); // Identifier does not exist on the other CircuitData
+                };
+
+                match (lhs_id_info, rhs_id_info) {
+                    (
+                        CircuitIdentifierInfo::Var(lhs_var_info),
+                        CircuitIdentifierInfo::Var(rhs_var_info),
+                    ) => {
+                        if lhs_var_info.get_type() != rhs_var_info.get_type()
+                            || !other_cd
+                                .vars
+                                .contains(self_cd.vars.get(lhs_var_info.get_var()).unwrap())
+                        {
+                            return Ok(false); // Not the same var type or UUID
+                        }
+                    }
+                    (
+                        CircuitIdentifierInfo::Stretch(lhs_stretch_info),
+                        CircuitIdentifierInfo::Stretch(rhs_stretch_info),
+                    ) => {
+                        if lhs_stretch_info.get_type() != rhs_stretch_info.get_type()
+                            || !other_cd.stretches.contains(
+                                self_cd
+                                    .stretches
+                                    .get(lhs_stretch_info.get_stretch())
+                                    .unwrap(),
+                            )
+                        {
+                            return Ok(false); // Not the same stretch type or UUID
+                        };
+
+                        // Check whether the declared stretches in the other CircuitData follow the same order of
+                        // declaration as in self. This is done by verifying that the indices of the declared stretches
+                        // in `identifier_info` of the other CircuitData - which match the stretches encountered during the
+                        // iteration here - are monotonically increasing.
+                        if let CircuitStretchType::Declare = rhs_stretch_info.get_type() {
+                            let rhs_stretch_idx =
+                                other_cd.identifier_info.get_index_of(id_name).unwrap();
```

Comment:
Is there a chance that the `unwrap()` call here might panic? We're getting information based on the `id_name` we extracted from `self_cd` into `other_cd`, how do we know that `id_name` exists in `other_cd` already?